### PR TITLE
fix the doc string and provide a clean, recursive

### PR DESCRIPTION
**Pull request by fieker**
version. The complex one actually does not work in all cases...
The problem was a weired ring where the base ring was not a field, but
did not have a base_ring anymore...

### CODE REVIEW

File: src/ResidueField.jl [hunk 1]
```diff
@@ -69,20 +69,12 @@ function modulus(r::ResFieldElem)
...
-   R = base_ring(r)
-   while R != Union{}
-      if typeof(R) <: Field
-         return characteristic(R)
-      end
-      R = base_ring(R)
-   end
-   return characteristic(base_ring(R))
+function characteristic(R::ResField)
+   return characteristic(base_ring(r))
```
Comment (thofma) [inline src/ResidueField.jl:77]:
```suggestion
   return characteristic(base_ring(R))
```

**Comment by wbhart** [comment]
This is ResidueField and so the characteristic function should return the characteristic of the field, not the modulus of the ResidueRing. We have modulus for returning the modulus. Probably the function is actually broken....

...perhaps your implementation is better.

**Comment by wbhart** [comment]
I'm not sure we can even compute this, except in special cases.

**Comment by fieker** [comment]
On Mon, Jul 26, 2021 at 01:25:30AM -0700, wbhart wrote:
> This is ResidueField and so the characteristic function should return the characteristic of the field, not the modulus of the ResidueRing. We have modulus for returning the modulus. Probably the function is actually broken....
> 
> ...perhaps your implementation is better.
> 
My implementation is mostly the same as the old one, just with different
termination condition.
In general, characteristic is difficult...

**Comment by wbhart** [comment]
It seems wrong for Z/pZ. It looks like it will return 0.

I guess I must have implemented the original, and not thought about what I was doing.

**Comment by thofma** [comment]
The original is also wrong for `ZZ/p`. There is a special implemenation a few lines below.

**Comment by wbhart** [comment]
Oh it's special cases. Ok.

**Comment by wbhart** [comment]
Oh, the problem here is I was reading the diff the wrong way around. You are removing the explicit iteration in favour of recursion and correcting the docstring.

Seems ok.

**Review by wbhart — APPROVED** [review]

**Pull request merged**